### PR TITLE
Fix issue with DoHandShakeMsgType/ShrinkInputBuffer when encryption i…

### PR DESCRIPTION
**Pull request by haydenroche5**
…s on (e.g.

during renegotiation).

This issue was brought to light by ZD 10911. When encryption is on (indicated
by the return value of IsEncryptionOn), DoHandShakeMsgType will finish up by
incrementing the input buffer index past the padding and MAC (if encrypt-then-
mac is enabled). In ProcessReply, if there are more messages to be read, the
index is decremented back before the padding and MAC. The issue arises when
ShrinkInputBuffer is called in between and copies data from the dynamic input
buffer to the static one. That function will get called with the index post-
increment, and thus the padding and MAC won't get copied into the static buffer,
which isn't what we want, since ProcessReply is going to decrement the index
since it thinks the padding and MAC are still there. This commit makes it so
the padding and MAC get included in the call to ShrinkInputBuffer when
encryption is on.

### CODE REVIEW

**Review by dgarske — APPROVED** [review]

**Comment by dgarske** [comment]
Over to @toddouska for final review.

File: src/internal.c [hunk 1]
```diff
@@ -12928,7 +12928,20 @@ static int DoHandShakeMsgType(WOLFSSL* ssl, byte* input, word32* inOutIdx,
...
     #endif
     ) {
-        ShrinkInputBuffer(ssl, NO_FORCED_FREE);
+        if (IsEncryptionOn(ssl, 0)) {
+            int extra = ssl->keys.padSz;
+        #if defined(HAVE_ENCRYPT_THEN_MAC) && !defined(WOLFSSL_AEAD_ONLY)
+            if (ssl->options.startedETMRead)
+                extra += MacSize(ssl);
+        #endif
+            ssl->buffers.inputBuffer.idx -= extra;
```
Comment (toddouska) [inline src/internal.c:12937]: Is this complete?  We need checks for idx >= extra.  What if idx -= extra allows size to just go down enough for static buffer use but then a couple lines later we in crease += extra that's bigger than the static buffer size.  How thoroughly has this been tested?

**Comment by dgarske** [comment]
ZD 11352 needs this. Please escalate is @haydenroche5

File: src/internal.c (hunk 1 above)
Comment (haydenroche5) [reply src/internal.c:12937]: I have only tested in using the case provided in ZD 10911. I will address your feedback now and push changes.

Comment (haydenroche5) [reply src/internal.c:12937]: I agree that the check to make sure extra isn't greater than the index is needed, though.

File: src/internal.c [hunk 2]
```diff
@@ -12950,7 +12950,24 @@ static int DoHandShakeMsgType(WOLFSSL* ssl, byte* input, word32* inOutIdx,
         && ssl->error != WC_PENDING_E && ssl->error != OCSP_WANT_READ
     #endif
     ) {
-        ShrinkInputBuffer(ssl, NO_FORCED_FREE);
+        if (IsEncryptionOn(ssl, 0)) {
+            word32 extra = ssl->keys.padSz;
+
+            if (extra > ssl->buffers.inputBuffer.idx)
```
Comment (haydenroche5) [inline src/internal.c:12956]: Added this check. Thanks Todd.

File: src/internal.c [hunk 3]
```diff
@@ -12950,7 +12950,25 @@ static int DoHandShakeMsgType(WOLFSSL* ssl, byte* input, word32* inOutIdx,
...
-        ShrinkInputBuffer(ssl, NO_FORCED_FREE);
+        if (IsEncryptionOn(ssl, 0)) {
+            word32 extra = ssl->keys.padSz;
+
+        #if defined(HAVE_ENCRYPT_THEN_MAC) && !defined(WOLFSSL_AEAD_ONLY)
+            if (ssl->options.startedETMRead)
+                extra += MacSize(ssl);
+        #endif
+
+            if (extra > ssl->buffers.inputBuffer.idx)
```
Comment (haydenroche5) [inline src/internal.c:12961]: Added this check. Thanks Todd.

**Comment by haydenroche5** [comment]
@SparkiDev David mentioned to me that you might have found another issue with this PR. Let me know if that's the case, and I can address it.

**Comment by dgarske** [comment]
> @SparkiDev David mentioned to me that you might have found another issue with this PR. Let me know if that's the case, and I can address it.

@SparkiDev told me "I think there could also be an issue that the cipher suite (MAC algorithm) changed.". @haydenroche5 if there is something please followup in a new PR.

**Review by dgarske — APPROVED** [review]
Reviewed and tested. Also fixes ZD 11352.

**Comment by SparkiDev** [comment]
Nope, it's fine. I was grasping at straws trying to find a difference and their code works with this patch.

**Pull request merged**